src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -572,8 +590,95 @@ internal static SymbolicRegexNode<S> MkConcat(SymbolicRegexBuilder<S> builder, S
             return concat;
         }
 
-        internal static SymbolicRegexNode<S> MkNot(SymbolicRegexBuilder<S> builder, SymbolicRegexNode<S> node) =>
-            Create(builder, SymbolicRegexKind.Not, node, null, -1, -1, default, null, SymbolicRegexInfo.Not(node._info));
+        internal static SymbolicRegexNode<S> MkNot(SymbolicRegexBuilder<S> builder, SymbolicRegexNode<S> root)
+        {
+            // Instead of just creating a negated root node
+            // Convert ~root to Negation Normal Form (NNF) by using deMorgan's laws and push ~ to the leaves
+            // This may avoid rather large overhead (such case was discovered with unit test PasswordSearchDual)
+            // Do this transformation in-line without recursion, to avoid any chance of deep recursion
+            // OBSERVE: NNF[node] represents the Negation Normal Form of ~node
+            Dictionary<SymbolicRegexNode<S>, SymbolicRegexNode<S>> NNF = new();
+            Stack<(SymbolicRegexNode<S>, bool)> todo = new();
+            todo.Push((root, false));
+            while (todo.Count > 0)
+            {
+                var top = todo.Pop();
+                bool secondTimePushed = top.Item2;
+                var node = top.Item1;
```

Comment:
Nit: we try to avoid using `var` unless the type is explicit on the right-hand side

Reply:
ok, will fix.